src/Agent.Sdk/Knob.cs

```diff
@@ -0,0 +1,134 @@
+            RuntimeVariableNames=new List<string>{ "AGENT_USE_NODE10" },
+            Description="Forces the agent to use Node 10 handler for all Node-based tasks",
+            DefaultValue="false"
+        };
+    }
+
+    public class KnobPanel
+    {
+        public class Knob : IKnob
+        {
+            public string Name { get; set; }
+            public List<string> EnvironmentVariableNames { get; set; }
+            public List<string> RuntimeVariableNames { get; set; }
+            // public TBDType CommandLineOption {get; private set;}
+            public string Description { get; set; }
+            public bool IsDeprecated {get; set; } = false;  // is going away at a future date
+            public bool IsExperimental {get; set; } = false; // may go away at a future date
+            public string DefaultValue {get; set; } = "";
+
+            public KnobValue GetValue(IKnobValueContext context)
+            {
+                //TODO: First thing we need to do is check if deprecated or experimental and throw an event that
+                //      can be logged
+
+                // proposed order of operations is
+                // 1. Run time variables
+                // 2. Command Line Options (Not Yet Implemented)
+                // 3. Environment Variables
+                // 4. Default
+                if (context != null)
+                {
+                    if (RuntimeVariableNames != null)
+                    {
+                        foreach (var runTimeVar in RuntimeVariableNames)
+                        {
+                            var value = context.GetVariableValueOrDefault(runTimeVar);
+                            if (value != null)
+                            {
+                                return new KnobValue(value, KnobSource.RuntimeVariable, runTimeVar);
+                            }
+                        }
+                    }
+                    if (EnvironmentVariableNames != null)
+                    {
+                        var scopedEnvironment = context.GetScopedEnvironment();
+                        foreach (var envVar in EnvironmentVariableNames)
+                        {
+                            var value = scopedEnvironment.GetEnvironmentVariable(envVar);
+                            if (!string.IsNullOrEmpty(value))
+                            {
+                                return new KnobValue(value, KnobSource.EnvironmentVariable, envVar);
+                            }
+                        }
+                    }
+                }
+                return new KnobValue(DefaultValue, KnobSource.BuiltinDefault);
+            }
+        }
+
+        public static ReadOnlyCollection<IKnob> GetAllKnobs()
+        {
+            List<IKnob> allKnobs = new List<IKnob>();
+            Type type = typeof(IKnob);
+            foreach (var info in type.GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly))
```

Comment:
This feels hacky, do we need to do it this way?

Reply:
I think there's a pattern that's easier to understand, checking it locally before I embarrass myself though :grin:

Reply:
```patch
diff --git a/src/Agent.Sdk/Knob.cs b/src/Agent.Sdk/Knob.cs
index ddfed272..76b661db 100644
--- a/src/Agent.Sdk/Knob.cs
+++ b/src/Agent.Sdk/Knob.cs
@@ -38,23 +38,19 @@ namespace Agent.Sdk
         }
     }
 
-    public class AgentKnobs : KnobPanel
+    public static class AgentKnobs
     {
-        public static readonly IKnob UseNode10 = new Knob()
+        public static readonly IKnob UseNode10 = new Knob("UseNode10")
         {
-            Name="UseNode10",
             EnvironmentVariableNames=new List<string>{ "AGENT_USE_NODE10" },
             RuntimeVariableNames=new List<string>{ "AGENT_USE_NODE10" },
             Description="Forces the agent to use Node 10 handler for all Node-based tasks",
             DefaultValue="false"
         };
-    }
 
-    public class KnobPanel
-    {
         public class Knob : IKnob
         {
-            public string Name { get; set; }
+            public string Name { get; private set; }
             public List<string> EnvironmentVariableNames { get; set; }
             public List<string> RuntimeVariableNames { get; set; }
             // public TBDType CommandLineOption {get; private set;}
@@ -63,6 +59,16 @@ namespace Agent.Sdk
             public bool IsExperimental {get; set; } = false; // may go away at a future date
             public string DefaultValue {get; set; } = "";
 
+            public Knob(string name)
+            {
+                Name = name;
+                if (AgentKnobs._knobs.ContainsKey(name))
+                {
+                    throw new ArgumentException($"Already have a knob called {name}", nameof(name));
+                }
+                AgentKnobs._knobs[name] = this;
+            }
+
             public KnobValue GetValue(IKnobValueContext context)
             {
                 //TODO: First thing we need to do is check if deprecated or experimental and throw an event that
@@ -105,20 +111,11 @@ namespace Agent.Sdk
 
         public static ReadOnlyCollection<IKnob> GetAllKnobs()
         {
-            List<IKnob> allKnobs = new List<IKnob>();
-            Type type = typeof(IKnob);
-            foreach (var info in type.GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly))
-            {
-                var instance = new Knob();
-                var locatedValue = info.GetValue(instance) as IKnob;
-
-                if (locatedValue != null)
-                {
-                    allKnobs.Add(locatedValue);
-                }
-            }
-            return allKnobs.AsReadOnly();
+            List<IKnob> knobs = new List<IKnob>(_knobs.Values);
+            return knobs.AsReadOnly();
         }
+
+        private static readonly IDictionary<string,IKnob> _knobs = new Dictionary<string,IKnob>();
     }
 
     public interface IKnob
```